P202.Training.Test/RoleServiceTest.cs

```diff
@@ -0,0 +1,128 @@
+﻿using AutoMapper;
+using Moq;
+using P202.Training.Data.Entities;
+using P202.Training.Data.Repositories;
+using P202.Training.Domain;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace P202.Training.Test
+{
+    public class RoleServiceTest
+    {
+        [Fact]
+        public void RoleService_CreateData()
+        {
+            // Arrange
+            Mock<IRoleRepository> roleRepository = new Mock<IRoleRepository>();
+            Mock<IMapper> mappingService = new Mock<IMapper>();
+
+            var newRole = new Role
+            {
+                Id = 1,
+                Name = "Admin"
+            };
+
+            roleRepository.Setup(r => r.CreateRole(newRole));           
+
+            mappingService.Setup(m => m.Map<Role, Role>(It.IsAny<Role>()))
+                            .Returns(newRole);          
+
+            // Act
+            roleRepository.Object.CreateRole(newRole);
+
+            var roleService = new RoleService(roleRepository.Object, mappingService.Object);
+            roleService.CreateRole(newRole);
+
+            //Assert that the Add method was called once
+            roleRepository.Verify(x => x.CreateRole(newRole), Times.Once());
+
+            Role expectedRole = roleService.ReadRole(1);
+
+            // Assert
+            Assert.Equal(newRole.Name, expectedRole.Name);
+
+        }
+
+        [Fact]
+        public void RoleService_DeleteData()
+        {
+            var deleteRole = new Role();
+            var mockRoleRepository = new Mock<IRoleRepository>();
+            mockRoleRepository.Setup(x => x.DeleteRole(deleteRole));
+            mockRoleRepository.Object.DeleteRole(deleteRole);
+
+            var mockMapper = new Mock<IMapper>();
+
+            var roleService = new RoleService(mockRoleRepository.Object, mockMapper.Object);
+            roleService.DeleteRole(deleteRole.Id);
+
+            mockRoleRepository.Verify(x => x.DeleteRole(deleteRole), Times.Once());
+        }
+
+        [Fact]
+        public void RoleService_UpdateData()
+        {
+            var updateRole = new Role { Name = "new-role-name" };
+            var mockRoleRepository = new Mock<IRoleRepository>();
+            mockRoleRepository.Setup(x => x.UpdateRole(updateRole)).Returns(updateRole);
+            mockRoleRepository.Object.UpdateRole(updateRole);
+
+            var mockMapper = new Mock<IMapper>();
+
+            var roleService = new RoleService(mockRoleRepository.Object, mockMapper.Object);
+            roleService.UpdateRole(updateRole.Id, updateRole);
+
+            mockRoleRepository.Verify(x => x.UpdateRole(updateRole), Times.Once());
+        }
+
+        [Fact]
+        public void RoleService_ReadData()
+        {
+            var role = new Role { Id = 3 };
+            var mockRoleRepository = new Mock<IRoleRepository>();
+            mockRoleRepository.Setup(x => x.GetRole(3)).Returns(role);
+            mockRoleRepository.Object.GetRole(3);
```

Comment:
If `GetRole` is not being used in the service because it's not implemented, what's the point of setting up the mock for the repository? It forces you to call the method in the test just to have something to assert. I think this whole test is moot until the method gets implemented for real.

Reply:
Done! I've implemented the service as you suggest